### PR TITLE
[Concurrency] Basic support for actor classes and actor isolation

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce the "actor class" syntax. Ensure that it is only used for
root classes or classes that inherit from other actor classes. Within
actors, implement a basic set of isolation rules for actors, which includes:

* Asynchronous actor methods can be invoked on any actor instance.
* All mutable instance properties and synchronous instance methods are
only accessible via "self" or "super" within the context of the actor.
* Closures defined within an actor method are considered to be a
context distinct from the actor itself, are therefore are subject to
the same restrictions as above.

Together with a scheduler that ensures that asynchronous invocations
for a particular actor are serialized (at least up until they hit a
suspension point or return), this serves to isolate the actor's
instance state in a shallow manner: references to shared mutable state
can still be passed among actors, code can access global variables,
and one can still form and pass around unsafe pointers.

Pair this with some basic checking within actor contexts that
identifies a few obvious potential kinds of race conditions:

* References to captured mutable local variables from within closures.
* References to mutable global and static/class variables.

The latter are warnings, for which we currently don't have a great
suppression mechanism. That, and a lot of tuning, are still to come.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: test/Concurrency/actor_isolation.swift [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    globalFunc()
+
+    // Class methods are okay.
+    Self.synchronousClass()
+    Self.synchronousStatic()
+
+    // Closures.
+    let localConstant = 17
+    var localVar = 17 // expected-note 2{{var declared here}}
+    acceptClosure {
```
Comment (atrick) [inline test/Concurrency/actor_isolation.swift:75]: The `acceptClosure` errors are way more restrictive than I thought. In the proposal, a function type must be declared `@concurrent` if it potentially crosses an actor boundary.

File: test/Concurrency/actor_isolation.swift [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      _ = text[0] // expected-warning{{actor-isolated property 'text' is unsafe to reference in code that may execute concurrently}}
+      _ = self.synchronous() // expected-warning{{actor-isolated instance method 'synchronous()' is unsafe to reference in code that may execute concurrently}}
+      _ = localVar // expected-warning{{local var 'localVar' is unsafe to reference in code that may execute concurrently}}
+      _ = localConstant
+    }
+
+    acceptEscapingClosure {
```
Comment (atrick) [inline test/Concurrency/actor_isolation.swift:82]: It's strange that `self.synchronous` is left out of the `acceptEscapingClosure` cases.

**Review by atrick — COMMENTED** [review]
I looked at this for the sake of understanding the isolation rules. If the rules are written down anywhere else, I'd love to know.

File: test/Concurrency/actor_isolation.swift (hunk 1 above)
Comment (ktoso) [reply test/Concurrency/actor_isolation.swift:75]: This surprised me as well, are not just escaping closures troublesome? To schedule a closure off somewhere I'd have to escape it, right?

**Review by ktoso — APPROVED** [review]
Fantastic stuff @DougGregor , minor comment/question inline :)

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 3]
```diff
@@ -161,3 +162,457 @@ bool IsAsyncHandlerRequest::evaluate(
...
+    ClassDecl *actorClass;
+  } data;
+
+  explicit IsolationRestriction(Kind kind) : kind(kind) { }
+
+  /// Determine whether the given value is an instance member of an actor
+  /// class that is isolated to the current actor instance.
+  ///
+  /// \returns the type of the actor.
+  static ClassDecl *getActorIsolatingInstanceMEmber(ValueDecl *value) {
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/Sema/TypeCheckConcurrency.cpp:241]: Thanks! Fixed locally

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 4]
```diff
@@ -161,3 +162,457 @@ bool IsAsyncHandlerRequest::evaluate(
...
+    /// e.g., because it's immutable.
+    Unrestricted,
+
+    /// Access to the declaration is unsafe in a concurrent context.
+    Unsafe,
+
+    /// The declaration can only be referenced from the given local context,
+    /// or a context that cannot execute concurrently with code in that
+    /// context.
+    Local,
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/Sema/TypeCheckConcurrency.cpp:216]: It's not really about actors, though. This is about captured locals being accessed in code that can execute concurrently. I'll rename to `LocalCapture` to make that more clear.

File: test/Concurrency/actor_isolation.swift [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    // Accesses on other actors can only reference immutable data or
+    // call asychronous methods
+    _ = otherActor.immutable // okay
+    _ = otherActor.synchronous() // expected-error{{actor-isolated instance method 'synchronous()' can only be referenced on 'self'}}
+    _ = await otherActor.asynchronous()
+    _ = otherActor.text[0] // expected-error{{actor-isolated property 'text' can only be referenced on 'self'}}
+
+    // Global data is okay if it is immutable.
+    _ = immutableGlobal
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply test/Concurrency/actor_isolation.swift:62]: There is a giant gaping hole here around reference types that I'm not going to touch in this particular pull request. We won't get away without adding type-level annotations, which need to be separately considered and introduced.

File: test/Concurrency/actor_isolation.swift (hunk 2 above)
Comment (DougGregor) [reply test/Concurrency/actor_isolation.swift:82]: Error of omission, I'll fix before merging. Thanks!

File: test/Concurrency/actor_isolation.swift (hunk 1 above)
Comment (DougGregor) [reply test/Concurrency/actor_isolation.swift:77]: The idea is to assume that non escaping functions won't be executed concurrently but escaping functions would be. I've now implemented that here. All of these rules are subject to change; this is very much an experiment.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci smoke test

**Comment by DougGregor** [comment]
Windows failure looks spurious:
```
fatal error C1083: Cannot open compiler generated file: 'T:\lldb\source\Plugins\Language\ObjC\CMakeFiles\lldbPluginObjCLanguage.dir\NSIndexPath.cpp.obj': Permission denied
```

File: test/Concurrency/actor_isolation.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (ktoso) [reply test/Concurrency/actor_isolation.swift:62]: Right, got it :)

File: lib/Sema/TypeCheckConcurrency.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Sema/TypeCheckConcurrency.cpp]: Ah I see, yeah that reads well, thanks :)

**Comment by davezarzycki** [comment]
If actors cannot inherit from classes and vice-versa, shouldn't they have a separate type within the type system? Or is this just implementation expedience?

**Pull request merged**